src/rt/region/region_rc.h

```diff
@@ -244,34 +503,40 @@ namespace verona::rt
         }
       }
 
-      while (!fin_q.empty())
+      reg->dealloc_objects(alloc, fin_q, sub_regions);
+    }
+
+    void dealloc_objects(
+      Alloc& alloc, ObjectStack& objects, ObjectStack& sub_regions)
+    {
+      while (!objects.empty())
       {
-        auto p = fin_q.pop();
-        RefCount* rc = p->get_ref_count();
-        reg->counts.remove(rc);
-        if (!p->is_trivial())
+        Object* o = objects.pop();
+        RefCount* rc = o->get_ref_count();
+        counts.remove(rc);
+        if (!o->is_trivial())
         {
-          p->finalise(in, sub_regions);
+          o->finalise(this, sub_regions);
         }
         // Unlike traced regions, we can deallocate this immediately after
```

Comment:
Why can you deallocate immediately.  If there are cycles, then finalisers could access into deallocated objects.

Reply:
Fixed here: 48271743cc6b0d0325250d3dc935045f04dfaf76